ENIGMAsystem/SHELL/Universal_System/Resources/AssetArray.h

```diff
@@ -0,0 +1,104 @@
+/** Copyright (C) 2019 Robert B. Colton
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#ifdef INCLUDED_FROM_SHELLMAIN
+#  error This file includes non-ENIGMA STL headers and should not be included from SHELLmain.
+#endif
+
+#ifndef E_ASSET_ARRAY
+#define E_ASSET_ARRAY
+
+#include <vector>
+#include <string>
+#include <utility>
+
+#ifdef DEBUG_MODE
+  #include "Widget_Systems/widgets_mandatory.h" // for show_error
+  #define CHECK_ID(id, ret) \
+    if (!exists(id)) { \
+      enigma_user::show_error("Requested " + asset_type + " asset " + std::to_string(id) + " does not exist.", false); \
+      return ret; \
+    }
+  #define CHECK_ID_V(id) CHECK_ID(id,)
+#else
+  #define CHECK_ID(id, ret)
+  #define CHECK_ID_V(id)
+#endif
+
+namespace enigma {
+
+// Asset storage container designed for dense cache-efficient resource processing.
+template<typename T>
+class AssetArray {
+public:
+  AssetArray(std::string asset_type): asset_type(asset_type) {}
+
+  int add(T asset) {
+    int id = assets.size();

```

Comment:
Here and throughout: this is `std::vector<T>::size_type`, or using `size_t` is fine with me.

Reply:
You don't seriously mean change the return type of add too do you? If you want to do that, then we should probably typedef that... the type used for resource ids. That way later this whole implementation could be swapped out for a slab allocator if somebody wanted to.

Reply:
You might be trying to have me walk off a cliff here actually. Some add functions in GM return -1 if there is an error loading them.

From the GM8.1 manual:
```
sprite_add_sprite(fname) Adds the sprite stored the file fname to the set of sprite resources. The file must be a .gmspr file that is saved in the sprite property form in GameMaker. As this file contains all sprite settings, no further arguments are required. The function returns the index of the new sprite that you can then use to draw it or to assign it to the variable sprite_index of an instance. When an error occurs -1 is returned.
```

Reply:
You can cast to an int for return. This is a compiler warning, as-written (integer truncation in assignment, comparison of signed vs unsigned).

Reply:
I see what you mean: a65ee6bf775ebf28f3381aa58bcc643c066c8af2

Though, I tried it both ways and couldn't get it to actually produce the warning at all. Regardless, later I want to typedef ids so the whole thing could potentially be swapped out with a custom slab allocator.